### ISSUE TITLE
Filtering with $null does not work on relational or media fields.

### ISSUE THREAD

**Issue by kocerb** [strapi/strapi#12225]
### Describe the bug
When try to filter by a relational field or media type field is null or not Strapi throws error: `Only $and, $or and $not can by used as root level operators. Found $null.`

### Steps to reproduce the behavior
1-) Create a collection type (let's say `foo`) that contains a relation with another collection type (let's say `bar`)
(or alternatively Create a collection type (let say `foo`) that contains a media type field (let's say `bar`))
2-) Send a **GET** request to the `/foos?populate=*&filters[bar][$null]=true`
3-) API returns 500 Internal Server Error and Strapi console logs the following message: 
> Only $and, $or and $not can by used as root level operators. Found $null.

### Expected behavior
On the earlier versions of Strapi (v3) there was support to filter by a relational field is null or not with a similiar query.

### System
Node.js version: v14.18.2
NPM version: v6.14.15
Strapi version: tried from v4.0.0 to v4.0.4 
Database: Postgresql 10
Operating system: macOS v11.6

**Comment by derrickmehaffy** [comment]
We should be able to filter on the related key of a relation (same happens if you try to filter for things like IDs without specifying the nested ID like:

**This Works**
```
filters: {
  someRelation: {
    id: {
      $in: [ 1,2,3 ]
    }
  }
}
```

**This Doesn't Work**
```
filters: {
  someRelation: {
    $in: [ 1,2,3 ]
  }
}
```

And it should

**Comment by thucltzigvy** [comment]
Any update? this is my query (strapi v4.0.7)
`count({
      where: {
        someRelation: {
          $notNull: true
        }
      }
    })`

**Comment by hennzen** [comment]
What I did instead of not-working
```
  filters: {
    $and: [
      {
        somerelation: {
          $notNull: true
        }
      }
    ]
  }
```
is
```
  filters: {
    $and: [
      {
        somerelation: {
          id: {
            $gte: 1
          }
        }
      }
    ]
  }
```

**Comment by jai2826** [comment]
This solution worked for me!!

I have a content type name "foo" and have another relational field name "bar"
foo has many to many relation with bar 
bar has only one field "name"
lets say name = [jack, joe ,jill] 
my fetch string is 
......../api/foos?filters[bar][name]=jill&populate=*
(ignore "&populate=*" if you don't know)
This will list and populate all the foos which have bar name as jill  

I hope it helps

**Comment by kraftdorian** [comment]
was able to get it working for the relation field type, using the notNull operator on the related content type fields:

```js
findContentType({
  filters: {
    relationFieldName: {
      id: {  // also works on the UID field type
        $notNull: true,
      },
    },
  },
})
```

using Strapi v4.3.4

**Comment by VaibhavPaliwal0007** [comment]
How can I combine notNull and my condition? 
<img width="294" alt="image" src="https://user-images.githubusercontent.com/78612380/187300254-817ac0db-29fe-4310-b0c7-abf9ca33509e.png">
Basically if the array is not empty then check for id

**Comment by exotexot** [comment]
This is still an issue for me on Strapi 4.7.0

I'm trying to check if there is valid relation to a media field (image hosted in mEdia Library), but it doesn't work



```
composedImage: {
              id: {
                $notNull: true,
              },
            },
```

still returns an element where:

```

{
    id: 392,
    createdAt: '2023-03-08T10:59:57.110Z',
    updatedAt: '2023-03-08T12:10:22.297Z',
    publishedAt: '2023-03-08T10:59:57.980Z',
    composedImage: null
  },
```

**Comment by derrickmehaffy** [comment]
@Marc-Roig this used to work in previous versions which is why it was classified as a bug and not an enhancement.

**Comment by loprima-l** [comment]
@derrickmehaffy Can we have some info, it's struggling point while using strapi ... that's a big lack of function as strapi doesn't support filtering on relation media columns@.

**Comment by derrickmehaffy** [comment]
> @derrickmehaffy Can we have some info, it's struggling point while using strapi ... that's a big lack of function as strapi doesn't support filtering on relation media columns.

Sorry, I don't really have any more information. The bug has been confirmed and assigned to the dev squads but I have absolutely zero control of when they pick up an issue to work on.

**Comment by exotexot** [comment]
If people need a workaround in the meantime they can write a custom controller for their endpoints. After querying their data, they will need to do the manual filtering by themselves using Array.filter()

**Comment by prawee** [comment]
> was able to get it working for the relation field type, using the notNull operator on the related content type fields:
> 
> ```js
> findContentType({
>   filters: {
>     relationFieldName: {
>       id: {  // also works on the UID field type
>         $notNull: true,
>       },
>     },
>   },
> })
> ```
> 
> using Strapi v4.3.4

thx! it's save my day

**Comment by oliverkidd** [comment]
For anyone reading this recently the above comment works well! Perfectly fine as a solution:

```
  filters: {
    relationFieldName: {
      id: {
        $notNull: true,
      },
    },
  },
```

**Comment by diacu** [comment]
Any idea how I can filter by media file? I just need a way to select all entities of a content where a field (media) is empty.. tried everything, but no solution yet.

**Comment by tabortrailways** [comment]
Has anyone made this workaround work via REST API? Every variation of $null/$notNull, true/false returns the full set of records in content for a url like the below.

`https://strapi.example.com/api/content?filters[relation][id][$null]=true`

**Comment by harrison-traintobecome** [comment]
> $gte



> Any idea how I can filter by media file? I just need a way to select all entities of a content where a field (media) is empty.. tried everything, but no solution yet.



> Any idea how I can filter by media file? I just need a way to select all entities of a content where a field (media) is empty.. tried everything, but no solution yet.

Anyone find the solution for querying where media is null/notNull

**Comment by BorisKamp** [comment]
Bump, need the filter for content types without a media field as well, something like this. I need to query all entries where a given media field is `null` in graphql. Right now the media field doesn't even pop up as an option in the graphql playground.

**Comment by alexandrebodin** [comment]
👋 this enhancement was merged and will be released in the upcoming 4.23.0 version.
